### ISSUE TITLE
Change Request: Allow shareable configs to apply only rules compatible with the active language

### ISSUE THREAD

**Issue by sindresorhus** [eslint/eslint#20999]
### ESLint version

v10.5.0

### What problem do you want to solve?

ESLint added useful validation for `meta.languages` recently. when a configured rule does not support the active language, ESLint throws. That is good for catching accidental misconfiguration.

However, this creates a difficult problem for shareable configs from plugins that contain rules for multiple languages.

For example, [`eslint-plugin-unicorn`](https://github.com/sindresorhus/eslint-plugin-unicorn) has mostly JavaScript rules, but also some rules that support CSS, HTML, JSON, Markdown, or any language. Ideally, users could write:

```js
export default defineConfig([
	{
		files: ['**/*.js'],
		extends: [unicorn.configs.recommended],
	},
	{
		files: ['**/*.json'],
		plugins: {json},
		language: 'json/json',
		extends: [unicorn.configs.recommended],
	},
]);
```

And ESLint would apply only the Unicorn rules whose `meta.languages` match the active language.

Currently, that config throws because the same `recommended` preset also contains JavaScript-only rules. The plugin can work around this by generating separate recommended configs per language, but that means every multi-language plugin has to duplicate ESLint's language matching logic and maintain multiple preset variants.

### What do you think is the correct solution?

Would you consider an optin way for shareable configs to filter incompatible rules by `meta.languages`?

For example, something like:

```js
{
	files: ['**/*.json'],
	plugins: {json},
	language: 'json/json',
	extends: [unicorn.configs.recommended],
	filterRulesByLanguage: true,
}
```

The current strict behavior could remain the default. The opt-in would let plugin presets "just work" in mixed-language configs when the correct language plugins are installed, while still preserving `meta.languages` as the source of truth.

Please focus on the problem statement and not too much on the proposed solution. There may be better solution for this.

### Participation

- [ ] I am willing to submit a pull request for this change.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

I also cannot just put `files` inside `unicorn.configs.recommended` since JS-only globs would stop the preset from working for TypeScript users, while TS/non-JS globs would cause ESLint to start matching files that may not have the required parser or language plugin configured...

**Comment by kirkwaiblinger** [comment]
relates closely to https://github.com/eslint/eslint/issues/20590 (problem statement is roughly the same, candidate solution different). I think this issue does a better job explaining how this is an issue for a multi-language shareable config, and that user-constructed globs as suggested in #20590 would be insufficient to address that use case (which I hadn't considered when authoring #20590).

**Comment by derekblackwood61-lang** [comment]
Thanks for linking #20590. I agree the underlying problem is essentially
the same, but this issue focuses specifically on the challenges faced by
multi-language shareable configs.
The new meta.languages validation is valuable because it catches
configuration mistakes early. However, it also means that shareable configs
containing rules for multiple languages can no longer be reused across
different language contexts without either:
splitting presets by language,
duplicating ESLint's language-selection logic inside plugins, or
forcing users to manually reconstruct presets.
The key requirement is not necessarily a specific option like
filterRulesByLanguage, but a mechanism that allows shareable configs to
declare a superset of rules and have ESLint automatically ignore rules
whose meta.languages do not match the active language.
That would preserve meta.languages as the single source of truth while
avoiding duplication across plugins and reducing maintenance burden for
multi-language ecosystems.
I think it's worth evaluating this as a config-composition problem rather
than a rule-validation problem, since the current strict behavior works
well for direct user configuration and the friction mainly appears when
composing reusable presets.

**Comment by nzakas** [comment]
This is actually working the way it should by design:
https://github.com/eslint/rfcs/pull/135#discussion_r2157016871

The consensus was that silently disabling rules was a bad user experience because users would never have a way of knowing that they were configuring a rule to run but it didn't.

**Comment by fregante** [comment]
> would never have a way of knowing

Doesn't `--print-config` do exactly this? If the config disables a rule for the selected file, then the rule wouldn't appear in the print config.

**Comment by sindresorhus** [comment]
@nzakas I had some time to think about this now. The RFC settled the default for rules a user configures directly, and I agree with it ther. But the design doc never mentions shareable configs, and a rule arriving through `extends` is a different case: the user did not name the rule, the preset did.

With `eslint-plugin-unicorn@74` and `eslint@10.9.1`:

```js
{
	files: ['**/*.json'],
	plugins: {json},
	language: 'json/json',
	extends: [unicorn.configs.recommended],
}
```

This throws with 304 rule names listed. Exactly 4 rules in that preset support `json/json`. To get those 4, a user has to write 304 `'off'` entries by hand and revisit the list every release.

The only workaround is per-language presets, and those are silent in both directions. A `recommendedJson` preset leaves the JS rules out, with no error and no way to ask ESLint why. `unicorn.configs.all` already does this with the CSS-only rules. Worse, a user extending `unicorn.configs.recommended` never finds out that unicorn has CSS, HTML, JSON, Markdown and YAML rules at all, so they silently miss rules that do apply to their files. That is the same failure the RFC wanted to avoid, just moved somewhere ESLint cannot report it.

It also does not scale. unicorn has three preset variants and rules for six languages, and every new language means more presets plus a config edit from every user before any of it reaches them. With language filtering, `extends: [unicorn.configs.recommended]` under each language block keeps working, and new language support arrives on upgrade.

The matching gets duplicated too, and gets it wrong. unicorn's own filter handles `'js/js'` and `'*'` and nothing else, so it drops any rule declaring `js/*`, which ESLint itself matches: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/9b6f37bc11e56bfbbe210f24baa95474bdd52a31/index.js#L51

And it pushes me to make `meta.languages` inaccurate. unicorn has a Markdown-only rule that declares `js/js` anyway, so `configs.all` stays loadable: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/9b6f37bc11e56bfbbe210f24baa95474bdd52a31/rules/require-frontmatter-fields.js#L90-L95 The check meant to keep `meta.languages` honest is what makes lying about it useful.

On users never knowing: an opt-in is not silent, the user typed it, and the rule still shows in `--print-config` as `[0]`, like anything else set to off.

`files` inside the preset does not work either. `files: ['**/*.js']` yields `{}` rules for someone scoping it to `**/*.ts`, and `files: ['**/*.json']` makes ESLint parse JSON with espree for anyone who has not configured a JSON language.

I'am not attached to `filterRulesByLanguage`, and #20590 may be the better vehicle. But every option today either duplicates ESLint's matching logic or makes the metadata wrong, which is not great.

Could you reopen this for more discussion? 🙏
